include/meshio/obj.hpp

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2017, Lakshman Anumolu, Pradeep Garigipati
+ * All rights reserved.
+ *
+ * This file is part of MeshIO whose distribution is governed by
+ * the BSD 2-Clause License contained in the accompanying LICENSE.txt
+ * file.
+ */
+
+#ifndef MESHIO_OBJ_OBJ_HPP_
+#define MESHIO_OBJ_OBJ_HPP_
+
+#define TINYOBJLOADER_IMPLEMENTATION
+
+#include "./utilities.hpp"
+#include "./meshio_defines.hpp"
+#include "../third_party/tiny_obj_loader.h"
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+
+namespace meshio
+{
+
+/* Datastructure to store Wavefront OBJ */
+template<class T>
+class OBJAttribute : public GeometryData<T> {
+  public:
+    std::string mOBJName;
+    std::vector<int> mMaterialIDs;
+
+    OBJAttribute() {}
+    ~OBJAttribute() {}
+
+    void resizeVertexAttributes(unsigned pNumTriangles) {
+        this->mPositions.resize(pNumTriangles);
+        this->mNormals.resize(pNumTriangles);
+        this->mTexturePositions.resize(pNumTriangles);
+        this->mColors.resize(pNumTriangles);
+        this->mParameterSpaceVertices.resize(pNumTriangles);
+    }
+
+    void resizeIndices(unsigned pNumIndices) {
+        this->mTriVertexIndices.resize(pNumIndices);
+        this->mTriTextureIndices.resize(pNumIndices);
+        this->mTriNormalIndices.resize(pNumIndices);
+    }
+
+    void clear() {
+        GeometryData<T>::clear();
+        mMaterialIDs.clear();
+    }
+
+    bool operator==(const OBJAttribute<T>& pOBJAttribute) const {
+        if (dynamic_cast<GeometryData<T>&>(
+                        const_cast<OBJAttribute<T>&>(*this)) ==
+                dynamic_cast<GeometryData<T>&>(
+                        const_cast<OBJAttribute<T>&>(pOBJAttribute))) {
+
+            if (not mOBJName.compare(pOBJAttribute.mOBJName)) {
+                // Checking equality for mMaterialIDs.
+                if (mMaterialIDs.size() && pOBJAttribute.mMaterialIDs.size()) {
+                    for (unsigned i = 0; i < mMaterialIDs.size(); ++i) {
+                        if (not (mMaterialIDs[i] == pOBJAttribute.mMaterialIDs[i])) {
+                            return false;
+                        }
+                    }
+                    return true;
+                } else if (not mMaterialIDs.size() &&
+                           not pOBJAttribute.mMaterialIDs.size()) {
+                    return true;
+                } else {
+                    return false;
+                }
+
+            } else {
+                return false;
+            }
+
+        } else {
+            return false;
+        }
+    }
+
+};
+
+template<class T>
+class OBJData {
```

Comment:
I think it is better to have an anonymous class defined and accessible only from OBJData instead of a class that is accessible from everywhere. It would go something like below

```cpp
class OBJData {
    typedef struct OBJAttributes {
        // ...
    } Attributes;
};
```
That way you can isolate tinyobjloader contents into a separate container class and also avoid creating an user-facing class - since meshio is header-only we have avoid anything and everything that can cause unintended usage by end users. In this case, they may try to use OBJAttributes. However, if the definition of this container is limited to the scope of OBJData class, then they can't even access it in the first place.